### ISSUE TITLE
feature request: impersonate users

### ISSUE THREAD

**Issue by konsultaner** [wamp-proto/wamp-proto#252]
We use our [connectanum](connectanum.com) implementation mostly for web applications or apps. Our clients were requesting remote control to those applications. The easist way for us to implement it would be impersonation. [Gitlab](https://about.gitlab.com/) uses this feature as well and we have implemented this with standard http software before.

Its not quite trivial to implement it into the wamp protocol. The easist way might be by utilizing the session meta api. Something like:

```
wamp.session.impersonate
```

and 

```
wamp.session.stop_impersonation
```

**Comment by oberstet** [comment]
To be honest, I neither do understand the use case, nor what "impersonating" WAMP sessions technically should mean. Could you explain a little bit more?

**Comment by konsultaner** [comment]
Impersonation could either mean that a session is passed from one channel A to another channel B. 
The router must then know that the channel A only imperonated the session and that there exist an original session that belongs to the channel A. Channel A that impersonated the other session from channel B acts as if it actually is channel B. Channel B is set to inactive, closed, some impersonated mode or the session is simply cloned.

OR

Impersonation could also mean that a channel has a session with role A and can apply for a second session with role B. The session B is the impersonated session. If session B is active the router knows about the two sessions. Session A is inactive for the channel while impersonation. Session B is the active session. If the channel stops impersonation the session a is active again. If the channel closes, both sessions are closed.

What is it for:
- maintaining - as gitlab uses it
- simple remote support

**Comment by oberstet** [comment]
There is no thing called a "channel" with WAMP.

**Comment by konsultaner** [comment]
I know, but I mean the SocketChannel with "channel" It is hard to explain without channel because it is neither the transport nor the session. It is the SocketChannel. But is it clear what I mean?

**Comment by oberstet** [comment]
No, sorry:(

There is no channel thing in Autobahn, Crossbar.io or WAMP .. so I assume it is an implementation artifact in Java?

Am 21.09.2016 2:21 nachm. schrieb Richard <notifications@github.com>:

I know, but I mean the SocketChannel with "channel" It is hard to explain without channel because it is neither the transport nor the session. It is the SocketChannel. But is it clear what I mean?

**Comment by konsultaner** [comment]
Thats right, it is a Java artifact. I'm sorry, but I am so used to it that I thought it was a normal thing. By channel I mean "open socket connection" that connects a router with a client.

**Comment by oberstet** [comment]
WAMP is not concerned about TCP sockets. Not event stream like transports (like pipes). It only talks about a reliable, bidirectional, ordered, message based tranport.

**Comment by konsultaner** [comment]
>WAMP is not concerned about TCP sockets.

Thats right, but its about transferring the wamp session to another connected user (socket) or the ability to apply for a second session. It is about the router definition.

**Comment by konsultaner** [comment]
This is how symfony explains impersonation:

>Sometimes, it's useful to be able to switch from one user to another without having to log out and log in again (for instance when you are debugging or trying to understand a bug a user sees that you can't reproduce).

**Comment by oberstet** [comment]
What I could imagine is for taking over a session is something like this:

* client A is connected and joined on a realm with session ID 1
* client B connects, and joins the same realm, sending a `HELLO` message with options `take_over_session=1`, authenticates, and if successful and authorized to take over sessions, the existing session is attached to client B and client A is disconnected at the same time

Nothing in above is about sockets, TCP or the like. The transports used by client A and B is irrelevant as far as session are concerned.

**Comment by konsultaner** [comment]
You are a lot better at describing things then me ;)
Yes, this is what I was trying to explain in version 2 after the "OR"

>Impersonation could also mean that a channel has a session with role A and can apply for a second session with role B. The session B is the impersonated session. If session B is active the router knows about the two sessions. Session A is inactive for the channel while impersonation. Session B is the active session. If the channel stops impersonation the session a is active again. If the channel closes, both sessions are closed.

The good thing about the `take_over_session` flag would be, that it works very well with the session meta api.

**Comment by oberstet** [comment]
Alright, so we are on the same sheet in principle, and above is something that could be well defined.

So I am reopening this for other implementors to comment.

> For me personally, this is very very low prio - but that isn't the benchmark of course. Let's see what other say.

**Comment by oberstet** [comment]
As mentioned, I understand the OP wish, and we have figured out a definition - however, without other chiming in, I tend to ... close;) for now, moving to backlog of "new features"

**Comment by konsultaner** [comment]
I seem to be the only one that needs this feature. I can live with a proprietary connectanum-router implementation, or would you like to see me contribute a definition here?

**Comment by gammazero** [comment]
This proposed feature needs to talk about authorization -- how is client B (taking over the session) allowed to have the things in the session as client A (the session creator).  For example, if A's authrole is authorized to create subscription X, and B's authrole is not, should B be allowed to take over A's session that has a subscription X?

If session takeover is in spec as described (https://github.com/wamp-proto/wamp-proto/issues/252#issuecomment-248649347) then session A should also specify that it will allow its session to be taken over, by sending a HELLO message with option `allow_take_over_session=1`.

This property is associated with the client not the session.  So, if session B also wants to allow the session to be taken over, B would also need to specify `allow_take_over_session=1`.

**Comment by konsultaner** [comment]
@gammazero The Idea behind impersonation is to let the admin takeover a session that is not accessable, because the secret is unknown. The admin should be able to see through a useres eyes.

> If session takeover is in spec as described (#252 (comment)) then session A should also specify that it will allow its session to be taken over, by sending a HELLO message with option allow_take_over_session=1

In my oppinion Session A should not be able to decide, especially because the router is the authority not a client.

> This proposed feature needs to talk about authorization -- how is client B (taking over the session) allowed to have the things in the session as client A (the session creator). For example, if A's authrole is authorized to create subscription X, and B's authrole is not, should B be allowed to take over A's session that has a subscription X?

I think client B inherits everithing that belongs to the session, such as Authroles, subscriptions, etc. I'm just not sure about registrations.

**Comment by ecorm** [comment]
I am working on a spec proposal for a WAMP-SCRAM authentication method (#135) . SCRAM inherits the SASL feature of an "authorization identifier" that is different than the user credentials. This feature allows essentially allows admins to "act" as other users.

I was going to leave the SASL "authorization identifier" out of the WAMP-SCRAM spec for simplicity, but I think I'll put it back in my proposal seeing that someone else might have a use for it.

**Comment by meejah** [comment]
Another thing to consider here is end-to-end encryption; an "admin" session couldn't possibly respond to or send RPCs or PubSubs in this case because it doesn't have the private keys (and it can never be permitted to get them, or there's no point in e2e encryption in the first place).

**Comment by ecorm** [comment]
> Another thing to consider here is end-to-end encryption; an "admin" session couldn't possibly respond to or send RPCs or PubSubs in this case because it doesn't have the private keys (and it can never be permitted to get them, or there's no point in e2e encryption in the first place).

The admin would use his own private keys for end-to-end encryption, but the application would behave as if a different user is logged in.

**Comment by gammazero** [comment]
@konsultaner OK, that makes sense given the use case:
> Session A should not be able to decide, especially because the router is the authority not a client.
Router can be configured to only allow sessions with certain authrole or authid to take over sessions.

I think would have to inherit registrations also, even if replied to their INVOCATIONS with an error.  Hopefully, the client implementations have some coordination and are able to handle what they are taking over.

That being said, the client doing the takeover will need to know subscription IDs and registration IDs of the session it is takes over, so that it knowns _what_ EVENT and INVOCATION it is receiving.  Is the client expected to use session meta to get this?

**Comment by konsultaner** [comment]
> I am working on a spec proposal for a WAMP-SCRAM authentication method (#135) . SCRAM inherits the SASL feature of an "authorization identifier" that is different than the user credentials. This feature essentially allows admins to "act" as other users.

Thinking of that, it is possible to have multiple sessions to the same authentication. To be honest it would be nice to have both SASL and Session takeover, but for my descriped described GitLab use case, SASL would be enough and would probably simplify a lot. @meejah  the end-to-end encription wouldn't be a problem anymore and (@gammazero ) subscriptions and registrations will most probably be applied by business logic after the authentication anyway.

**Comment by ecorm** [comment]
To clarify, the SASL/SCRAM "authorization identifier" would simply allow an admin "Alice" to **login as** a different user "Bob" using Alice's credentials. If Bob was already logged in, both Alice and Bob are running separate sessions. It would be as if Bob logged into the same web app on a second browser tab, but Alice is in control of the second tab. If Bob makes changes in the first tab, Alice may need to refresh her tab if the application was not designed to publish those changes.

**Comment by ecorm** [comment]
What I'm proposing is impersonating _users_, not "impersonating sessions". We can either rename this issue to "impersonate users", or create a new issue for a "impersonate users" feature.

The idea of an "authorization identifier" different from the username is not specific to SCRAM. The same idea could be used for all other authentication methods.

**Comment by meejah** [comment]
> The admin would use his own private keys for end-to-end encryption, but the application would behave as if a different user is logged in.

For e2e encryption, the keys are distributed out-of-band so I can't imagine how the above would work. 

Let's look at a concrete example: component A registers an endpoint `foo` and component B calls it. When calling, B is encrypting for A's public key -- so the router/take-over session can't decrypt the arguments (it doesn't have A's private key for this endpoint). The reply must be encrypted to B's public-key, so again the router can't decrypt the reply (it also doesn't have B's private key).

Assuming the router had grabbed the public keys (somehow), it could encrypt a reply to B but B should reject this as it is not signed by A (I'd have to double-check how NaCL implements this, but "authenticated encryption" should mean crypto_box is using signatures).

**Comment by ecorm** [comment]
> For e2e encryption, the keys are distributed out-of-band so I can't imagine how the above would work.

I suppose the admin would need to have already set-up his own keys and use those for e2e encryption. The admin uses his own keys, but tells the router/application that he's acting as a different user.  I admit that a separate set of encryption keys for an admin might not be practical/feasible in a customer support scenario.

In a customer support scenario involving e2e encryption, perhaps some kind of temporary key/token could be generated and sent out-of-band to the CSR. The CSR then authenticates remotely with that temporary key/token.

**Comment by meejah** [comment]
"uses his own keys" just won't work; the data is `crypto_box`-ed for a particular endpoint (e.g. "component A").

Certainly if this is "just" a distributed application and the use-case is debugging and the person operating the router is also operating and running all the components then the "admin" session/debugging-component could be given access to ALL the keys -- but it would still have to use the right ones (i.e. the ones the components are expecting). However, in such a case I don't see why you'd be using e2e crypto anyway?

The real use-case for using e2e encryption between RPC/PubSub components is when you *don't* trust the router -- so either this feature is just not available then or some other scheme would need to be employed. A really simple way would be to simply turn off e2e crypto (since you'd have to re-register the RPCs with new encryption keys anyway -- might as well just turn it off) and then use the "session takeover" feature as-is.

**Comment by meejah** [comment]
Just to clarify the above in case it's not obvious, by "e2e crypto" I'm talking about the crossbar feature employing NaCL crypto_box, mentioned here: https://github.com/wamp-proto/wamp-proto/issues/229

**Comment by ecorm** [comment]
For SCRAM authentication, the `HELLO` message would look something like:

```javascript
    [1, "realm1",
        {
            "roles": ...,
            "authmethods": ["wamp-scram-sha-256"],
            "authid": "bob",
            "username": "admin",
            "nonce": "rOprNGfwEbeRWgbNEkqO"
        }
    ]
```

The router would check against the "admin" password, but would use "bob" as the authid for that session. The rest of the application would treat the session as if it were "bob" that was authenticated.

Of course the router (or dynamic authenticator) would need to determine if "admin" has permission to act as a different user, but that would be a router configuration detail that's outside the scope of the WAMP protocol.

**Comment by ecorm** [comment]
> Just to clarify the above in case it's not obvious, by "e2e crypto" I'm talking about the crossbar feature employing NaCL crypto_box, mentioned here: #229 where the "ends" are components.

Oh I see. I assumed the "ends" in e2e crypto was between the client and the router, presumably over an unsecure transport. Thanks for the clarification. Yes I could see how admin crypto keys wouldn't work in that scenario.

**Comment by konsultaner** [comment]
However, in such a case I don't see why you'd be using e2e crypto anyway?

@meejah If the router is provided as a hosted service the router is not trusted, but may have an admin that might want to impersonate a user.

**Comment by konsultaner** [comment]
@meejah I have not implemented e2e crypto yet, but as far as I understand what you wrote, the same problems should appear if a user has authenticated more then once to that same router, right?

**Comment by meejah** [comment]
@konsultaner if you don't trust the router, you definitely don't want it to have the power to impersonate a user :)

I have a blog post that describes an e2e encryption setup (it's not JUST that though, so .. be warned :) here: https://meejah.ca/blog/end-to-end-encrypted-rpc-over-tor

There is TLS encryption between components and the router; this is transport encryption (and of course the router can read all data). The "e2e" encryption I mentioned above is for encrypting the payloads of messages. This is independent of any transport encryption. Once connected, components authenticate to the router; this is also independent of e2e encryption (but of course, components can't start passing messages until they are authenticated to the router). Perhaps this diagram (from the above blog post) helps clarify:

![diagram](https://user-images.githubusercontent.com/145599/36570705-855bf186-17f0-11e8-9607-05970e6ef379.png)

**Comment by konsultaner** [comment]
@meejah 
> if you don't trust the router, you definitely don't want it to have the power to impersonate a user :)

The router is only not trusted in seeing what is delivered in the payload. Imagine a project for a hostpital or other medical environment. As a service provider you would like to make sure that your service (router) is not able to read any payload. This makes contracts belonging the privacy a lot easier and and make the service more trustable. The service provider still has the major authority and so has the router.

**Comment by konsultaner** [comment]
> I suppose the admin would need to have already set-up his own keys and use those for e2e encryption.

If a user A impersonates another user B, nothing happens to user B's sessions. User A receives its own session. So user A has to goodbye his own session if logged in and login with the SASL feature. The router would then store user A's and user B's public key. User A has the same authId as user B and the same roles. It is pretty much the same scenario as if user B logs in twice?

If a user publishes to a topic that has been subscibed by A or B. Both receive an event including the public key. The same happens with an invocation. If the registered procedures are in single mode and not in round robin or something else. The router answers with an error in both scenarios.

> Certainly if this is "just" a distributed application and the use-case is debugging and the person operating the router is also operating and running all the components then the "admin" session/debugging-component could be given access to ALL the keys 

@meejah Aren't those keys are just the public keys anyway?

**Comment by oberstet** [comment]
fwiw, I have no clue what @konsultaner is talking about;)

**Comment by oberstet** [comment]
the discussion here is about e2e, not about the OPs feature request of session impersonation, for which there does not seem lots of interest. so closing  --  if anyone disagrees, pls vote for the OPs original feature request (deliberate, controlled WAMP session impersonation .. which is _not_ what "impersonation" refers to when talking about an attack scenario on e2e)

**Comment by meejah** [comment]
> Aren't those keys are just the public keys anyway?

No, these are completely separate keypairs. To send data you need the private part. They have nothing to do with authentication (or SASL). The keys themselves are never sent over any WAMP connection (distributed out-of-band). There could be dozens of these keypairs (i.e. up to one pair for each RPC registration, for just A). These are also separate from the identity keypair for the component.

**Comment by konsultaner** [comment]
@meejah ok, I see! It would still be nice to see SASL in the final SCRAM definition. Even if this means that e2e is impossible.
@oberstet closing this issue makes sense since SASL would be the solution. by the way it had been changed from "impersonate session" to "impersonate user".

**Comment by ecorm** [comment]
Sorry, my bad for suggesting that the issue be renamed to "impersonate user. The original discussion was indeed about impersonating sessions.
